```{r, include=FALSE}
library(ggplot2)
library(RColorBrewer)
library(scales)
library(knitr)
library(VennDiagram) ## to plot Venn diagram
library(DT)

#gene_meta[sample_id, condition]
#gene.de[significant == "yes",  contrast, `log2(fold_change)`, gene_id, gene, p_value, 
#        q_value, avgFPKM]

knitr::opts_chunk$set(fig.height = 4*gene.de[, .N, by = contrast][, .N])
```


```{r}

intgroup.dt <- gene_meta[, .(sample_id, condition)][order(sample_id)]
intgroup.dt[, color_luminance := 100-(10*1:.N), by = condition]
intgroup.dt <- merge(intgroup.dt, 
                     intgroup.dt[, .(samples_per_condition = .N), 
                                 by = condition][, hue := brewer.pal(8, 
                                                                     "Set2")[1:length(condition)]][], 
                     by = "condition")[, color := muted(hue, l = color_luminance)][]
```

# Summary

```{r summary samples conditions}
datatable(data = gene_meta[, .(sample_id, condition)],
          rownames = F, 
          style = "bootstrap", class = "compact display", 
          caption = "Samples and their condition", 
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))

rm(.Random.seed)
set.seed(1)
```

The analysis considered `r gene_meta[, .N, by = sample_id][, .N]` samples grouped in `r gene_meta[, .N, by = condition][, .N]` conditions.  

```{r summary samples per condition}
datatable(data = gene_meta[, .(samples = .N), by = condition],
          rownames = F, 
          style = "bootstrap", class = "compact display", 
          caption = "Number of replicates for each condition group", 
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))
```

The contrasts considered were `r paste0(unique(gene.de[, contrast]), collapse = ", ")`.  

# Differential expression test counts 

The number of significantly differential expressed genes (DEG) is reported in the table below.  

```{r summary de genes per condition}
kable(x = gene.de[significant == "yes", .(`Tot. DEGs` = .N), by = contrast], 
      row.names = F, 
      caption = "Number of differentially expressed genes for each contrast")
```

```{r MA plot, fig.cap="N.B: failed differential expression tests are not shown in the plot", dev='jpeg'}
ma.plot <- ggplot(data = gene.de[!is.infinite(`log2(fold_change)`),], 
                  mapping = aes(x = avgFPKM, y = `log2(fold_change)`, 
                                color = significant)) + 
    geom_point(size = 1, alpha = 0.5) + scale_x_log10()
if(gene.de[, .N, by =contrast][, .N] > 1){
    ma.plot <- ma.plot + facet_grid(facets = contrast ~ . , scales = "free")
}
ma.plot + xlab("Average FPKM") + ylab(expression(Log[2](FC))) +
    scale_color_manual(name = "Differentially expressed", 
                       values = c("black", "red"), na.value = "gray") + 
    theme(legend.position="bottom")
```

The number of DEGs shared among the contrasts is reported below.  

```{r DEG shared}
## de tab
shared.counts.table <- dcast(gene.de[significant == "yes", ], 
                             formula = gene_id ~ contrast, 
                             value.var = "significant")[, -c("gene_id"), 
                                                        with = F]
kable(x = crossprod(!is.na(shared.counts.table)), row.names = T, 
      caption = "Number of differentially expressed genes shared among contrasts")
```

```{r}
## de venn
n.contrasts <- length(unique(gene.de[, contrast]))
if(n.contrasts > 1){
    fill.colors <- 1:n.contrasts
    if(n.contrasts < 5){fill.colors <- fill.colors +1}
    if(n.contrasts < 6){
        dec.lists.by.contrast <- split(gene.de[significant == "yes", gene_id],
                                       f = gene.de[significant == "yes", contrast])
        #venn(deg.lists.by.contrast) ## venn using gplots
        vp <- venn.diagram(dec.lists.by.contrast,
                           fill = fill.colors,
                           alpha = 0.3, filename = NULL,
                           main = "Differentially expressed genes shared by the contrasts",
                           cex = 3, main.cex = 2, cat.cex = 1.5)
        grid.draw(vp)
    }else{
        print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
    }
}
```

## Differentially expressed gene table

```{r DEG table}
gene.de.tab <- dcast(gene.de[, 
        .(gene_id = gene_id, gene, contrast, `log2(fold_change)`, p_value, 
          q_value, avgFPKM)], 
      formula = gene_id + gene ~ contrast, 
      value.var = c("log2(fold_change)", "avgFPKM", "q_value", "p_value"))

## save table
gene.de.tab.filename <- file.path(results.dir, "DEG_tests_by_contrast.csv")
write.csv(gene.de.tab, file = gene.de.tab.filename, row.names = F)
```

DEG list (by contrast) was saved in file `r gene.de.tab.filename`

```{r show DEG table}
## display only DEGs
setkey(gene.de.tab, gene_id)
datatable(data = gene.de.tab[unique(gene.de[significant == "yes", .(gene_id)]),], 
          rownames = F, 
              style = "bootstrap", class = "compact display", 
              caption = "Differential tests of gene expression. Only DEGs in at least one contrast are reported", 
              fillContainer = F, autoHideNavigation = T, filter = "top")
```

# Significant DEGs shared between the contrasts

## Over-expressed genes

```{r up DEG venn}
## up tab 
sign.up <- gene.de[significant == "yes" & `log2(fold_change)` >= 0, ]
if(nrow(sign.up) > 0 ){
    up.shared.counts.table <- dcast(sign.up, 
                                    formula = gene_id ~ contrast, 
                                    value.var = "significant")[, -c("gene_id"), 
                                                               with = F]
    
    datatable(data = crossprod(!is.na(up.shared.counts.table)), 
              rownames = T,
              style = "bootstrap", class = "compact display", 
              caption = "Number of over expressed genes shared among contrasts",
              fillContainer = F, autoHideNavigation = T, options = list(searching = F))
    
    ## up venn
    n.contrasts <- length(unique(gene.de[, contrast]))
    if(n.contrasts > 1){
        fill.colors <- 1:n.contrasts
        if(n.contrasts < 5){fill.colors <- fill.colors +1}
        if(n.contrasts < 6){
            deg.lists.by.contrast <- split(gene.de[significant == "yes" & `log2(fold_change)` >= 0,
                                                   gene_id], 
                                           f = gene.de[significant == "yes" & `log2(fold_change)` >= 0,
                                                       contrast])
            #venn(deg.lists.by.contrast) ## venn using gplots
            vp <- venn.diagram(deg.lists.by.contrast, 
                               fill = fill.colors, 
                               alpha = 0.3, filename = NULL,
                               main = "Over-expressed genes shared by the contrasts", 
                               cex = 3, main.cex = 2, cat.cex = 1.5)
            grid.draw(vp)
        }else{
            print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
        }
    }else{
        print("Only one contrast")
    }
}else{
    print("No genes overexpressed")
}
```

## Lower-expressed genes

```{r down DEG venn}
## down tab 
sign.down <- gene.de[significant == "yes" & `log2(fold_change)` < 0, ]
if(nrow(sign.down) > 0 ){
    down.shared.counts.table <- dcast(sign.down, 
                                      formula = gene_id ~ contrast, 
                                      value.var = "significant")[, -c("gene_id"), 
                                                                 with = F]
    
    datatable(data = crossprod(!is.na(down.shared.counts.table)), 
              rownames = T,
              style = "bootstrap", class = "compact display", 
              caption = "Number of lower expressed genes shared among contrasts",
              fillContainer = F, autoHideNavigation = T, options = list(searching = F))
    
    ## down venn
    n.contrasts <- length(unique(gene.de[, contrast]))
    if(n.contrasts > 1){
        fill.colors <- 1:n.contrasts
        if(n.contrasts < 5){fill.colors <- fill.colors +1}
        if(n.contrasts < 6){
            deg.lists.by.contrast <- split(sign.down[, gene_id],
                                           f = sign.down[, contrast])
            #venn(deg.lists.by.contrast) ## venn using gplots
            vp <- venn.diagram(deg.lists.by.contrast, 
                               fill = fill.colors, 
                               alpha = 0.3, filename = NULL,
                               main = "Lower-expressed genes shared by the contrasts", 
                               cex = 3, main.cex = 2, cat.cex = 1.5)
            grid.draw(vp)
        }else{
            print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
        }
    }else{
        print("Only one contrast")
    }
}else{
    print("No genes lower expressed")
}
```


# Session info
```{r sessionInfo}
sessionInfo()
```