---
title: "CirComPara: circular RNA differential expression"
date: "`r Sys.Date()`"  
output: 
  html_document: 
    keep_md: no
    number_sections: yes
    toc: yes
    toc_depth: 2
    toc_float: 
      collapsed: false
      smooth_scroll: true
    theme: "readable"
---

```{r global_options, include=FALSE}
## variables to be defined in parent frame
# results.dir
# circrnas.per.sample.file
# meta.file
fdr.thr <- 0.1

dir.create(path = results.dir, showWarnings = F, recursive = T)
knitr::opts_chunk$set(fig.width = 12, 
                      fig.height = 6, 
                      dev = c('svglite', 'png'), 
                      fig.pos = "center",
                      fig.path = file.path(results.dir, paste0('Figs_de', .Platform$file.sep)),
                      echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE)

knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = results.dir)

suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(VennDiagram))
```

```{r load data, echo=FALSE, include=FALSE}
non.value.columns <- c("circ_id", "gene_ids", "gene_names", 
                       "label", "chr", "start", "end", "strand")
circrnas.per.sample <- fread(circrnas.per.sample.file)
circrnas.per.sample.matrix <- as.matrix(round(circrnas.per.sample[, -non.value.columns, 
                                                                  with = FALSE]))
rownames(circrnas.per.sample.matrix) <- circrnas.per.sample$circ_id

sampleTable <- unique(read.csv(file = meta.file)[, c("sample", "condition")])
rownames(sampleTable) <- as.character(sampleTable[, "sample"])
countData <- circrnas.per.sample.matrix

ddsHTSeq <- DESeqDataSetFromMatrix(countData = countData[, row.names(sampleTable)],
                                   colData = sampleTable,
                                   design = ~ condition)
dds <- DESeq(ddsHTSeq)
## add additional feature data
# featureData <- data.frame(gene=rownames(countData))
# (mcols(dds) <- DataFrame(mcols(dds), featureData))
```

# Summary

Groups cosidered: `r paste0(unique(sampleTable$condition), collapse = ", ")`  

```{r samples and condition}
datatable(data = as.data.table(sampleTable)[, .(sample, condition)],
          rownames = F, 
          style = "bootstrap", class = "compact display", 
          caption = "Samples and their condition", 
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))
```

The analysis considered `r as.data.table(sampleTable)[, .N, by = sample][, .N]` samples grouped in `r as.data.table(sampleTable)[, .N, by = condition][, .N]` conditions.  

```{r number of samples per condition}
datatable(data = as.data.table(sampleTable)[, .(samples = .N), by = condition],
          rownames = F, 
          style = "bootstrap", class = "compact display", 
          caption = "Number of replicates for each condition group", 
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))
```

# Principal component analysis (PCA) considering circRNA expression
```{r pca, echo=FALSE}
## PCA plot
#rld <- rlog(dds, blind = T)
vsl <- vst(dds, blind = T)
pca <- prcomp(t(assay(vsl)))

suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggthemes))

intgroup <- c("condition")
percentVar <- pca$sdev^2/sum(pca$sdev^2)
## compute PC variance explained % and plot
percentVar.df <- data.frame(PCs = paste0("PC", 1:length(percentVar)), 
                            Variation_percentage = percentVar*100)
percentVar.df$PCs <- factor(percentVar.df$PCs, levels = percentVar.df$PCs, ordered = T)
pc.var.explained.plot <- ggplot(data = percentVar.df, aes(x = PCs, y = Variation_percentage)) + 
    geom_bar(stat = "identity") + 
    geom_text(aes(label = round(Variation_percentage, 2)), vjust = -.5) + 
    ylab("Variation explained %") + theme_bw() + 
    ggtitle("Amount of variation explained by the principal components") + 
    theme(plot.title = element_text(hjust = 0.5))

## compute and plot PCs
intgroup.df <- as.data.frame(colData(vsl)[, intgroup, drop = FALSE])
group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
} else {
    colData(vsl)[[intgroup]]
}
d <- data.frame(pca$x, group = group, intgroup.df, name = colnames(vsl))
pca.plot <- ggplot(data = d, aes(PC1, PC2, color=condition, shape=condition, label=name)) + 
    geom_point(size = 4, color = "black") + geom_point(size = 3, alpha = 1) + coord_fixed() + 
    geom_text(position="jitter", hjust=0.2, vjust=-0.2, size=6, show.legend = F) +
    xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
    ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance"))+ 
    scale_color_brewer(palette = "Set2") + scale_x_continuous(expand = c(0.3,0)) +
    scale_y_continuous(expand = c(0.3,0)) + theme_bw() + 
    theme(legend.position="bottom") + 
    ggtitle("Samples plotted according to the first two principal components") + 
    theme(plot.title = element_text(hjust = 0.5))
```

```{r plot pca, echo=FALSE}
pc.var.explained.plot
pca.plot
```

# Differential expression test counts 

```{r differential expression, echo=FALSE}
# resultsNames(dds)
conditions <- contrasts(sampleTable$condition, contrasts = T)
res.contrasts <- list()
res.contrast.names <- c()
for(i in 1:(dim(conditions)[1]-1)){
    for(j in 1:dim(conditions)[2])
        if(conditions[i,j] == 0){
            contrast.name <- paste(rownames(conditions)[i], "vs", colnames(conditions)[j])
            res <- results(dds, 
                           contrast = c("condition", rownames(conditions)[i], colnames(conditions)[j]), 
                           alpha = fdr.thr)
            #res$circ_id <- rownames(res)
            res.contrast.names <- c(res.contrast.names, contrast.name)
            res.contrasts <- c(res.contrasts, res)
        }
}
names(res.contrasts) <- res.contrast.names
```

The contrasts considered were: `r paste0(res.contrast.names, collapse = ", ")`.  

```{r differentially expressed circRNA counts, include=FALSE}
test.results.dt <- rbindlist(lapply(res.contrasts, 
                                    function(x)as.data.table(as.data.frame(x, row.names = rownames(x)), 
                                                             keep.rownames = T)), 
                             idcol = "contrast")
colnames(test.results.dt)[colnames(test.results.dt) == "rn"] <- "circ_id"

test.results.dt[, significant := ifelse(padj <= fdr.thr, "yes", "no")]
# res.contrast.summary <- function(x){
#     print(names(x))
#     print(summary(x[[1]]))
# }
# sapply(res.contrasts, res.contrast.summary)
#sapply(res.contrasts, summary)
```

The number of significantly (FDR <= `r fdr.thr`) differential expressed circRNAs (DECs) is reported in the table below.  

```{r summary de genes per condition}
datatable(data = test.results.dt[significant == "yes", .(`Tot. DECs` = .N), by = contrast],
          rownames = F, 
          style = "bootstrap", class = "compact display", 
          caption = "Number of differentially expressed circRNAs for each contrast",
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))
```


```{r MA plot, fig.height=12, dev='jpeg'}
ma.plot <- ggplot(data = test.results.dt, 
                  mapping = aes(x = baseMean, y = log2FoldChange, 
                                color = significant)) + 
    geom_point(size = 1, alpha = 0.5) + scale_x_log10()
if(length(res.contrast.names) > 1){
    ma.plot <- ma.plot + facet_grid(facets = contrast ~ . , scales = "free")
}
ma.plot + xlab("Average RPM") + ylab(expression(Log[2](FC))) +
    scale_color_manual(name = "Differentially expressed", 
                       values = c("black", "red"), na.value = "gray") + 
    theme(legend.position="bottom")
```

The number of DECs shared among the contrasts is reported below.  

```{r DECs shared}
shared.counts.table <- dcast(test.results.dt[significant == "yes", ],
                             formula = circ_id ~ contrast,
                             value.var = "significant")[, -c("circ_id"),
                                                        with = F]

datatable(data = crossprod(!is.na(shared.counts.table)), 
          rownames = T,
          style = "bootstrap", class = "compact display", 
          caption = "Number of differentially expressed circRNAs shared among contrasts",
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))

n.contrasts <- length(unique(test.results.dt[, contrast]))
if(n.contrasts > 1){
    fill.colors <- 1:n.contrasts
    if(n.contrasts < 5){fill.colors <- fill.colors +1}
    if(n.contrasts < 6){
    dec.lists.by.contrast <- split(test.results.dt[significant == "yes", circ_id],
                                   f = test.results.dt[significant == "yes", contrast])
    #venn(deg.lists.by.contrast) ## venn using gplots
    vp <- venn.diagram(dec.lists.by.contrast,
                       fill = fill.colors,
                       alpha = 0.3, filename = NULL,
                       main = "Differentially expressed circRNAs shared by the contrasts",
                       cex = 3, main.cex = 2, cat.cex = 1.5)
    grid.draw(vp)
    }else{
        print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
    }
}
```

## Over-expressed circRNAs

```{r up DECs venn}
## up tab 
up.shared.counts.table <- dcast(test.results.dt[significant == "yes" & log2FoldChange >= 0, ],
                             formula = circ_id ~ contrast,
                             value.var = "significant")[, -c("circ_id"),
                                                        with = F]

datatable(data = crossprod(!is.na(up.shared.counts.table)), 
          rownames = T,
          style = "bootstrap", class = "compact display", 
          caption = "Number of over expressed circRNAs shared among contrasts",
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))

## up venn
n.contrasts <- length(unique(test.results.dt[, contrast]))
if(n.contrasts > 1){
    fill.colors <- 1:n.contrasts
    if(n.contrasts < 5){fill.colors <- fill.colors +1}
    if(n.contrasts < 6){
    dec.lists.by.contrast <- split(test.results.dt[significant == "yes" & log2FoldChange >= 0, circ_id],
                                   f = test.results.dt[significant == "yes" & log2FoldChange >= 0, contrast])
    #venn(deg.lists.by.contrast) ## venn using gplots
    vp <- venn.diagram(dec.lists.by.contrast,
                       fill = fill.colors,
                       alpha = 0.3, filename = NULL,
                       main = "Over-expressed circRNAs shared by the contrasts",
                       cex = 3, main.cex = 2, cat.cex = 1.5)
    grid.draw(vp)
    }else{
        print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
    }
}
```

## Lower-expressed circRNAs

```{r down DECs venn}
## down tab
down.shared.counts.table <- dcast(test.results.dt[significant == "yes" & log2FoldChange < 0, ],
                             formula = circ_id ~ contrast,
                             value.var = "significant")[, -c("circ_id"),
                                                        with = F]

datatable(data = crossprod(!is.na(down.shared.counts.table)), 
          rownames = T,
          style = "bootstrap", class = "compact display", 
          caption = "Number of lower expressed circRNAs shared among contrasts",
          fillContainer = F, autoHideNavigation = T, options = list(searching = F))

## down venn
n.contrasts <- length(unique(test.results.dt[, contrast]))
if(n.contrasts > 1){
    fill.colors <- 1:n.contrasts
    if(n.contrasts < 5){fill.colors <- fill.colors +1}
    if(n.contrasts < 6){
    dec.lists.by.contrast <- split(test.results.dt[significant == "yes" & log2FoldChange < 0, circ_id],
                                   f = test.results.dt[significant == "yes" & log2FoldChange < 0, contrast])
    #venn(deg.lists.by.contrast) ## venn using gplots
    vp <- venn.diagram(dec.lists.by.contrast,
                       fill = fill.colors,
                       alpha = 0.3, filename = NULL,
                       main = "Lower-expressed circRNAs shared by the contrasts",
                       cex = 3, main.cex = 2, cat.cex = 1.5)
    grid.draw(vp)
    }else{
        print("Too many contrasts to be plotted by a Venn diagram. Skipped.")
    }
}
```


## Differentially expressed circRNA table

```{r show differential expression test, echo=FALSE}
res.data.table <- dcast(test.results.dt, formula = circ_id ~ contrast, value.var = c("baseMean", "log2FoldChange",
                                                              "lfcSE", "stat", "pvalue", "padj"))
# res.data.table <- as.data.table(as.data.frame(res.contrasts), keep.rownames = T)
# colnames(res.data.table)[colnames(res.data.table) == "rn"] <- "circ_id"

res.data.table.annotated <- merge(res.data.table, 
                                  circrnas.per.sample[, non.value.columns, with = F], 
                                  by = "circ_id")
## save table
circrna.diff.exp.table.file <- file.path(dirname(circrnas.per.sample.file), "circrna_differential_expression_tests.csv")
write.csv(x = res.data.table.annotated, file = circrna.diff.exp.table.file)

## show table
padj.columns <- colnames(res.data.table.annotated)[grepl(pattern = "padj", 
                                                            x = colnames(res.data.table.annotated))]
res.data.table.annotated$any.de <- as.logical(rowSums(res.data.table.annotated[, lapply(.SD, function(x){x <= .1}), 
                                                                    .SDcols=padj.columns], na.rm = T))

numeric.columns <- colnames(res.data.table.annotated)[grepl(pattern = "baseMean|log2FoldChange|lfcSE|stat|pvalue|padj", 
                                                            x = colnames(res.data.table.annotated))]
columns.to.show <- colnames(res.data.table.annotated)[
    grepl(pattern = "circ_id|log2FoldChange|padj|gene_ids|gene_names|label", 
          x = colnames(res.data.table.annotated))]
basemean.columns <- colnames(res.data.table.annotated)[grepl(pattern = "baseMean", x = colnames(res.data.table.annotated))]
res.data.table.annotated$baseMean <- round(rowMeans(res.data.table.annotated[, basemean.columns, with = F]), 2)
DT::datatable(res.data.table.annotated[, (numeric.columns) := round(.SD, 2), 
                                       .SDcols=numeric.columns][any.de == TRUE, ][, c(columns.to.show, "baseMean"), with = F], 
              rownames = F, 
              style = "bootstrap", class = "compact display", 
              caption = "CircRNA differential expression tests. Missing circRNAs failed the statistical test.", 
              fillContainer = F, autoHideNavigation = T, filter = "top")

```

Full results of differential expression tests can be found in file `r circrna.diff.exp.table.file`

```{r other plots, echo=FALSE}
#plotMA(res, main="DESeq2", ylim=c(-2,2))
# d <- plotCounts(dds, gene=which.min(test.results.dt$padj), intgroup="condition", returnData=TRUE)
# ggplot(d, aes(x=condition, y=count)) +
#     geom_boxplot() + geom_point(position=position_jitter(w=0.1,h=0)) +
#     scale_y_log10()
```

# Differentially expressed circRNA loci

```{r DE circrna loci, echo=FALSE}
de.circ.chr.hist <- melt(dcast(test.results.dt[padj <= .1, 
                                               .(chr = sub(pattern = "(^[^:]*):.*", replacement = "\\1", x = circ_id)), 
                                               by = .(contrast, circ_id)][
                                                   , .N, by = .(contrast, chr)], 
                               formula = chr ~ contrast, fill = 0, value.var = "N"), 
                         variable.name = "contrast", value.name = "count", 
                         id.vars = "chr")
de.circ.chr.hist$chr <- factor(de.circ.chr.hist$chr, levels = c(1:22, "X", "Y", "MT"), ordered =TRUE)
ggplot(de.circ.chr.hist, aes(x = chr, y = count, group = contrast, fill = contrast)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    ggtitle("Amount of differentially expressed circRNAs per chromosome") + 
    theme(plot.title = element_text(hjust = 0.5))
```

# Session info
```{r}
sessionInfo()
```
