---
title: "CirComPara: gene expression analysis"
date: "`r Sys.Date()`"  
output: 
  html_document: 
    keep_md: no
    number_sections: yes
    toc: yes
    toc_depth: 2
    toc_float: 
      collapsed: false
      smooth_scroll: true
    theme: "readable"
---

```{r global_options, include=FALSE}
## variables to be defined in parent frame
# results.dir <- "analysis/cuffdiff"
# meta.file <- "analysis/cuffdiff/read_groups.info"
# gene.xpr.file <- "analysis/cuffdiff/genes.read_group_tracking" # if cufflinks. Comma separated list of *gene_abund.tab files if stringtie
# sample # comma separated list if multisample
# linear.method # comma separated list if more than one method. Currently use only the first in the list
# gene_raw_counts_list <- "" #Comma separated list
# trx_raw_counts_list <- "" #Comma separated list

dir.create(path = results.dir, showWarnings = F, recursive = T)
knitr::opts_chunk$set(fig.width = 12, 
                      fig.height = 6, 
                      dev = c('svglite', 'png'), 
                      fig.pos = "center",
                      fig.path = file.path(results.dir, paste0('Figs_geneexp', .Platform$file.sep)),
                      echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE)

knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = results.dir)

library(data.table)
library(ggplot2)
library(RColorBrewer)
library(scales)
library(DESeq2)
```

```{r}
##prepare meta file
if(basename(meta.file) == "read_groups.info"){
    gene_meta <- fread(meta.file, colClasses = c("character", "factor", "character", 
                                                 "numeric", "numeric", "numeric", 
                                                 "numeric"))
    gene_meta[, sample_id := sub(".*cuffquant/(.*)_hisat2.*", "\\1", file)]
}else{
    ## prepare meta file and sample colors for plots
    gene_meta <- unique(fread(file = meta.file)[, .(sample_id = sample, condition)])
}
intgroup.dt <- gene_meta[, .(sample_id, condition)][order(sample_id)]
intgroup.dt[, color_luminance := 100-(10*1:.N), by = condition]
intgroup.dt <- merge(intgroup.dt, 
                     intgroup.dt[, .(samples_per_condition = .N), 
                                 by = condition][, hue := brewer.pal(8, 
                                                                     "Set2")[1:length(condition)]][], 
                     by = "condition")[, color := muted(hue, l = color_luminance)][]
```


```{r load data, include=FALSE}
gene.xpr.file           <- readLines(gene.xpr.file)
transcripts.gtf.files   <- readLines(transcripts.gtf.files)
gene_raw_counts_list    <- readLines(gene_raw_counts_list)
trx_raw_counts_list     <- readLines(trx_raw_counts_list)

## use only the first method in the list
method <- strsplit(linear.method, ",")[[1]][1]
if(method == "cufflinks"){
    trx_raw_counts <- fread(trx_raw_counts_list)
    
    if(basename(meta.file) == "read_groups.info"){
        genes.read_group_tracking <- 
            fread(gene.xpr.file, 
                  colClasses = c("character", "factor", "character", 
                                 "integer", "numeric", "numeric", 
                                 "numeric", "character", "character"))[
                                     , .(tracking_id, condition, replicate, raw_frags, internal_scaled_frags, FPKM)]
        
        setkey(gene_meta, condition, replicate_num)
        setkey(genes.read_group_tracking, condition, replicate)
        gene.raw.frags.dt <- 
            dcast(genes.read_group_tracking[gene_meta[, .(sample_id, condition, 
                                                          replicate_num)]][, .(gene_id = tracking_id, 
                                                                               sample_id, raw_frags)], 
                  formula = gene_id ~ sample_id, value.var = "raw_frags", fill = 0)
        
        ## save gene_raw_counts
        ## https://github.com/cole-trapnell-lab/cufflinks/issues/12
        ## simply multiply each samples count table by the library size factors to "de-normalize" the counts
        gene_raw_counts <- genes.read_group_tracking
        gene_raw_counts <- gene_raw_counts[, `:=`(raw.reads = ceiling(raw_frags),
                                                  gene_id = tracking_id)]
        gene_raw_counts <- merge(gene_raw_counts, 
                                gene_meta[, .(condition, replicate_num, sample_id)], 
                                by.x = c("condition", "replicate"), 
                                by.y = c("condition", "replicate_num"))
        
        message("Will use 'raw_frags' from genes/isoforms.read_group_tracking files as raw counts")
        write.csv(x = dcast(gene_raw_counts,
                        formula = gene_id ~ sample_id,
                        value.var = "raw.reads", fill = 0),
                  file = file.path(results.dir, "gene_expression_rawcounts_table.csv"),
                  row.names = F)

        ## save trx_raw_counts
        trx_raw_counts$replicate <- as.character(trx_raw_counts$replicate)
        trx_raw_counts[, raw.reads := ceiling(raw_frags)]
        trx_raw_counts <- merge(trx_raw_counts, 
                                gene_meta[, .(condition, replicate_num, sample_id)], 
                                by.x = c("condition", "replicate"), 
                                by.y = c("condition", "replicate_num"))
        
        trx_raw_counts.dt <- dcast(trx_raw_counts[, .(transcript_id = tracking_id, sample_id, raw.reads)],
                            formula = transcript_id ~ sample_id,
                            value.var = "raw.reads", fill = 0)
        write.csv(x = trx_raw_counts.dt,
                  file = file.path(results.dir, "transcript_expression_rawcounts_table.csv"), 
                  row.names = F)
        
    }else{
        # tracking_id, class_code, nearest_ref_id, 
        # gene_id, gene_short_name, tss_id, 
        # locus, length, coverage, 
        # FPKM, FPKM_conf_lo, FPKM_conf_hi, FPKM_status
        genes.read_group_tracking <- fread(gene.xpr.file, 
                                           colClasses = c("character", "character", "character", 
                                                          "character", "character", "character",
                                                          "character", "character", "numeric",
                                                          "numeric", "numeric", "numeric", "character"))[
                                                              , .(tracking_id, FPKM, coverage)]
        genes.read_group_tracking$sample_id <- unique(gene_meta$sample_id)
        
        message("Will use 'coverage' from genes/isoforms.read_group_tracking files as raw counts")
        ## TODO: should coverage be multiplied by gene/transcript length and divided by 
        ## average read length to get raw counts? (as for StringTie)
        
        ## save trx_raw_counts
        write.csv(x = trx_raw_counts[, .(transcript_id = tracking_id, raw.reads = ceiling(coverage))],
                  file = file.path(results.dir, "transcript_expression_rawcounts_table.csv"), 
                  row.names = F)
        
        ## save gene_raw_counts: gene coverage as the sum of transcript coverage
        write.csv(x = trx_raw_counts[, .(raw.reads = ceiling(sum(coverage))), by = gene_id],
                  file = file.path(results.dir, "gene_expression_rawcounts_table.csv"),
                  row.names = F)
    }
}
if(method == "stringtie"){
    gene.xpr.file.paths <- strsplit(gene.xpr.file, ",")[[1]]
    gene.xpr.list <- lapply(gene.xpr.file.paths, fread)
    names(gene.xpr.list) <- sub("_gene_abund.tab", "", 
                                sapply(gene.xpr.file.paths, basename, 
                                       USE.NAMES = F))
    gene.xpr <- rbindlist(gene.xpr.list, use.names = T, idcol = "sample")

    gene.xpr.tpm <- dcast(data = gene.xpr,
                          formula = `Gene ID` + `Gene Name` + Reference + Strand + Start + End ~ sample,
                          value.var = "TPM", fill = 0)
    write.csv(x = gene.xpr.tpm, 
              file = file.path(results.dir, "gene_expression_TPM_table.csv"), 
              row.names = F)
    gene.xpr.nreads <- dcast(data = gene.xpr, 
                             formula = `Gene ID` + `Gene Name` + Reference + Strand + Start + End ~ sample, 
                             value.var = "Coverage", fill = 0)
    write.csv(x = gene.xpr.nreads, 
              file = file.path(results.dir, "gene_expression_Nreads_table.csv"), 
              row.names = F)
    
    gene.raw.frags.dt <- gene.xpr.nreads[, c("Gene ID", gene_meta$sample_id), 
                                         with = F][, `:=`(gene_id = `Gene ID`, `Gene ID` = NULL)][]
    genes.read_group_tracking <- gene.xpr[, .(sample_id = sample, tracking_id = `Gene ID`, FPKM)]
    
    
    ## merge gene_raw_counts_list
    gene_raw_counts_list.paths <- strsplit(gene_raw_counts_list, ",")[[1]]
    gene_raw_counts <- lapply(gene_raw_counts_list.paths, fread)
    names(gene_raw_counts) <- sub("_gene_expression_rawcounts.csv", "",
                                              sapply(gene_raw_counts_list.paths, basename,
                                                     USE.NAMES = F))
    gene_raw_counts <- rbindlist(gene_raw_counts, use.names = T, idcol = "sample_id")
    
    write.csv(x = dcast(gene_raw_counts,
                        formula = gene_id ~ sample_id,
                        value.var = "raw.reads", fill = 0),
              file = file.path(results.dir, "gene_expression_rawcounts_table.csv"),
              row.names = F)

    ## merge trx_raw_counts_list
    trx_raw_counts_list.paths <- strsplit(trx_raw_counts_list, ",")[[1]]
    trx_raw_counts <- lapply(trx_raw_counts_list.paths, fread)
    names(trx_raw_counts) <- sub("_transcript_expression_rawcounts.csv", "",
                                              sapply(trx_raw_counts_list.paths, basename,
                                                     USE.NAMES = F))
    trx_raw_counts <- rbindlist(trx_raw_counts, use.names = T, idcol = "sample_id")
    
    write.csv(x = dcast(trx_raw_counts[raw.reads > 0, ],
                        formula = gene_id + transcript_id ~ sample_id,
                        value.var = "raw.reads", fill = 0),
              file = file.path(results.dir, "transcript_expression_rawcounts_table.csv"), 
              row.names = F)
}
```

```{r}
if(method == "htseq_count"){
    gene.xpr.file.paths <- strsplit(gene.xpr.file, ",")[[1]]
    
    sampleTable <- gene_meta
    sampleTable$file <- sapply(sampleTable$sample_id, grep, 
                               x = gene.xpr.file.paths, value = T, fixed = T)
    
    dds <- DESeqDataSetFromHTSeqCount(sampleTable = as.data.frame(sampleTable[, .(sample_id, file, condition)]),
                                      directory = "",
                                      design = ~ 1)
    dds <- estimateSizeFactors(dds)
    
    ## write gene raw counts
    write.csv(x = as.data.table(counts(dds), keep.rownames = "gene_id"),
              file = file.path(results.dir, "gene_expression_rawcounts_table.csv"),
              row.names = F)

    ## TODO: transcript raw counts 
    message("Transcript raw counts yet to be implemented with htseq-count")
    trx_raw_counts <- matrix()
    write.csv(x = trx_raw_counts,
              file = file.path(results.dir, "transcript_expression_rawcounts_table.csv"), 
              row.names = F)
}
```


```{r save gene expression in table format}
if(meta.file != ""){
    if(method == "cufflinks"){
        expressed_genes <- genes.read_group_tracking[, FPKM := round(FPKM, digits = 8)][FPKM > 0]
        if(basename(meta.file) == "read_groups.info"){
            setkey(gene_meta, condition, replicate_num)
            setkey(expressed_genes, condition, replicate)
            expressed_genes <- expressed_genes[gene_meta[, .(sample_id, condition, replicate_num)]]
        }
        gene.xpr.fpkm <- dcast(expressed_genes[, .(gene = tracking_id, sample_id, FPKM)], 
                               formula = gene ~ sample_id, 
                               value.var = "FPKM", 
                               fill = 0)
    }
    if(method == "stringtie"){
        expressed_genes <- genes.read_group_tracking[, FPKM := round(FPKM, digits = 8)][FPKM > 0]
        
        gene.xpr.fpkm <- dcast(data = gene.xpr,
                               formula = `Gene ID` + `Gene Name` + Reference + Strand + Start + End ~ sample,
                               value.var = "FPKM", 
                               fill = 0)[, gene := `Gene ID`][, c("gene", gene_meta$sample_id), with = F]
        
        expressed_genes <- merge(expressed_genes, gene_meta, by = "sample_id")
    }
    if(method == "htseq_count"){
        ## normalized expression
        gene.xpr.fpkm <- as.data.table(counts(dds, normalized = T), keep.rownames = "gene")
        warning("Using DESeq2 normalized values instead of FPKM since htseq-count was selected.")
        
        expressed_genes <- merge(melt(gene.xpr.fpkm, 
                                      id.vars = "gene", 
                                      variable.name = "sample_id", 
                                      value.name = "FPKM")[
                                          FPKM > 0, .(sample_id, tracking_id = gene, FPKM)], 
                                 gene_meta, by = "sample_id")
    }
}
expressed_genes_table.file <- file.path(results.dir, "gene_expression_FPKM_table.csv")
write.csv(x = gene.xpr.fpkm, file = expressed_genes_table.file, row.names = F)
```

Expression of genes was saved in tabular format in file <a href="`r basename(expressed_genes_table.file)`">`r basename(expressed_genes_table.file)`</a>.  

# Principal component analysis (PCA) considering gene expression

```{r compute pca, echo=FALSE}
if(meta.file != "" & length(gene_meta$sample_id)>1){
    non.value.columns <- c("gene")
    
    genes.per.sample.matrix <- as.matrix(gene.xpr.fpkm[, -non.value.columns, with = FALSE])
    rownames(genes.per.sample.matrix) <- gene.xpr.fpkm$gene
    
    sampleTable <- as.data.frame(gene_meta[, .(sample = sample_id, condition)])
    rownames(sampleTable) <- as.character(sampleTable[, "sample"])
    
    # remove not expressed genes
    countData <- genes.per.sample.matrix[rowSums(genes.per.sample.matrix) > 0, ]
    
    pca.plot <- tryCatch({
        
        do.scale <- ifelse(ncol(countData) > 2, T, F)
        
        pca <- prcomp(t(countData), scale. = do.scale)
        
        intgroup <- c("condition")
        percentVar <- pca$sdev^2/sum(pca$sdev^2)
        ## compute PC variance explained % and plot
        percentVar.df <- data.frame(PCs = paste0("PC", 1:length(percentVar)), 
                                    Variation_percentage = percentVar*100)
        percentVar.df$PCs <- factor(percentVar.df$PCs, levels = percentVar.df$PCs, ordered = T)
        pc.var.explained.plot <- ggplot(data = percentVar.df, aes(x = PCs, y = Variation_percentage)) + 
            geom_bar(stat = "identity") + 
            geom_text(aes(label = round(Variation_percentage, 2)), vjust = -.5) + 
            ylab("Variation explained %") + theme_bw() + 
            ggtitle("Amount of variation explained by the principal components") + 
            theme(plot.title = element_text(hjust = 0.5))
        
        ## compute and plot PCs
        # intgroup.df <- as.data.frame(colData(rld)[, intgroup, drop = FALSE])
        intgroup.df <- data.frame(intgroup.dt[, c("sample_id", intgroup), with = F], row.names = "sample_id")
        group <- if (length(intgroup) > 1) {
            factor(apply(intgroup.df, 1, paste, collapse = " : "))
        } else {
            #colData(rld)[[intgroup]]
            intgroup
        }
        
        d <- data.frame(pca$x, group = group, 
                        condition = intgroup.df[colnames(countData), ], 
                        name = colnames(countData))
        pca.plot <- ggplot(data = d, aes(PC1, PC2, color=condition, shape=condition, label=name)) + 
            geom_point(size = 4, color = "black") + geom_point(size = 3, alpha = 1) + coord_fixed() + 
            geom_text(position="jitter", hjust=0.2, vjust=-0.2, size=6, show.legend = F) +
            xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
            ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance"))+ 
            scale_color_brewer(palette = "Set2") + scale_x_continuous(expand = c(0.3,0)) +
            scale_y_continuous(expand = c(0.3,0)) + theme_bw() + 
            theme(legend.position="bottom") + 
            ggtitle("Samples plotted according to the first two principal components") + 
            theme(plot.title = element_text(hjust = 0.5))
        
        list(pca.plot = pca.plot, pc.var.explained.plot = pc.var.explained.plot)
    }, 
    error = function(cond){
        return(NULL)
    }
    )
}

```

```{r plot pca, echo=FALSE}
if(meta.file != "" & length(gene_meta$sample_id)>1){
    
    if(!is.null(pca.plot)){
        print(pca.plot$pc.var.explained.plot)
        print(pca.plot$pca.plot)
    }else{
        print("Failed to compute principal components")
    }
}else{
    print("PCA not applicable to single sample")
}
```

# Gene expressed by sample

```{r show_gene_expressed_by_sample}
expressed_genes.by.sample <- expressed_genes[, .N, 
                                             by = .(sample_id, 
                                                    tracking_id)][, .("Tot. gene expressed" = .N), 
                                                                  by = sample_id]
knitr::kable(rbind(expressed_genes.by.sample[order(sample_id),], list("Non-redundant", 
                                            expressed_genes[, .N, by = tracking_id][, .N])))
```

```{r plot_gene_expressed_by_sample}
default.font.size <- 12
expressed_genes.by.sample$sample_id <- factor(expressed_genes.by.sample$sample_id, 
                                                 levels = gene_meta[order(condition, sample_id), sample_id], 
                                                 ordered = T)
expressed_genes.by.sample.plot <- ggplot(data = expressed_genes.by.sample, 
                                         mapping = aes(x = sample_id, y = `Tot. gene expressed`,
                                                       fill = sample_id)) + 
  geom_bar(stat = "identity", width = .7, colour="black") + 
  geom_text(stat='identity', aes(label = `Tot. gene expressed`), vjust=-0.3, size=default.font.size*(5/14)) +
  ggtitle("Gene expressed per sample") + xlab("") + ylab("Number of genes") + 
  theme_bw()
## compute limits to remove space from x-axis. Add a 10% increase to the top limit
expressed_genes.by.sample.plot.limits <- c(0, max(expressed_genes.by.sample$`Tot. gene expressed`)*1.1)
expressed_genes.by.sample.plot <- expressed_genes.by.sample.plot + guides(fill = FALSE) +
  theme(text = element_text(size=default.font.size), axis.text.x = element_text(angle=45, hjust=1)) + 
  scale_y_continuous(expand = c(0, 0), limits = expressed_genes.by.sample.plot.limits) + 
        theme(plot.title = element_text(hjust = 0.5))

if(meta.file != ""){
    expressed_genes.by.sample.plot + 
        scale_fill_manual(values = setNames(intgroup.dt[, color], nm = intgroup.dt[, sample_id]))
}else{
    expressed_genes.by.sample.plot
}
```

# Gene expression distribution by sample

```{r show_gene_expression}
knitr::kable(sapply(X = split(expressed_genes, expressed_genes$sample_id), 
       FUN = function(x){summary(x$FPKM)}), caption = "Gene expression (FPKM) distribution")
```

```{r plot_gene_expression}
## boxplot expression per sample: genes
expressed_genes$sample_id <- factor(expressed_genes$sample_id, 
                                                 levels = gene_meta[order(condition, sample_id), sample_id], 
                                                 ordered = T)

gene.xpr.per.sample.boxplot <- ggplot(data = expressed_genes, 
                                      mapping = aes(x = sample_id, y = FPKM, fill = sample_id)) + 
  geom_boxplot(notch = T, varwidth = T) + scale_y_log10(labels = scales::comma) + xlab(NULL) + ylab("FPKM") + 
  ggtitle("Gene expression\nper sample") + theme_bw()
gene.xpr.per.sample.boxplot <- gene.xpr.per.sample.boxplot + guides(fill=FALSE) + 
  theme(text = element_text(size=default.font.size), axis.text.x = element_text(angle=45, hjust=1)) + 
        theme(plot.title = element_text(hjust = 0.5))
if(meta.file != ""){
    gene.xpr.per.sample.boxplot <- gene.xpr.per.sample.boxplot + 
        scale_fill_manual(values = setNames(intgroup.dt[, color], nm = intgroup.dt[, sample_id]))
}
gene.xpr.per.sample.boxplot
```

# Cumulative expression

```{r cumulative_expression, include=FALSE}
cum_gene_by_sample <- expressed_genes[order(sample_id, -FPKM), 
                                      .(csum = cumsum(FPKM), seqid = seq_along(.I), tracking_id), 
                                      by = .(sample_id)]
cum_gene_by_sample[, cond_tot := max(csum), by = .(sample_id)][, perc_csum := csum/cond_tot]

if(meta.file != ""){
    cum_gene_by_sample <- merge(cum_gene_by_sample, intgroup.dt[, .(sample_id, condition)], by = "sample_id")
}else{
    cum_gene_by_sample$condition <- "unset"
}
```

```{r cumulative_expression_box, fig.width=12, fig.height=6}
## Cumulative expression plot: genes
gene.cumsum.per.sample <- ggplot(data = cum_gene_by_sample, mapping = aes(x = seqid, y = perc_csum, 
                                                                          color = sample_id, linetype = condition)) +
  geom_line() + ggtitle("Gene cumulative expression") + xlab("Number of genes") +
  ylab("Cumulative expression") + scale_y_continuous(labels = scales::percent) +
  theme_bw()
gene.cumsum.per.sample <- gene.cumsum.per.sample +
  theme(text = element_text(size=default.font.size), legend.position = "bottom") + 
        theme(plot.title = element_text(hjust = 0.5))

if(meta.file != ""){
    gene.cumsum.per.sample <- gene.cumsum.per.sample + 
        scale_color_manual(values = setNames(intgroup.dt[, color], nm = intgroup.dt[, sample_id]), "Sample") + 
        guides(color = guide_legend(ncol = intgroup.dt[, .N, by = condition][, .N]), 
               linetype = guide_legend(direction = "vertical"))
}else{
    gene.cumsum.per.sample <- gene.cumsum.per.sample + scale_color_discrete("Sample")
}
gene.cumsum.per.sample
```

# Session info
```{r sessionInfo}
sessionInfo()
```
